### DISCUSSION TITLE
Javascript sign transaction request() problems

### DISCUSSION THREAD

**Discussion by Virus-Axel** [phantom/docs#118] [Help & Support] [answered]
### Summary

Hi! I am developing a Solana SDK for godot. I am unable to create correct signatures from phantom in javascript. I would really like some details on exactly what phantom does to its input and what it expects as input to sign properly.

### Steps to Reproduce

When trying to sign transactions from javascript. I am limited to using the request method. I read in the example at [https://docs.phantom.app/solana/sending-a-transaction](https://docs.phantom.app/solana/sending-a-transaction) that phantom expects `message: bs58.encode(transaction.serializeMessage())` as parameters. This seems odd and I am pretty sure I am supposed to serialize the whole transaction with empty signature. Anyways here is what I get:

![Screenshot from 2023-11-08 08-08-40](https://github.com/phantom/docs/assets/14315050/23f09ed3-df6a-4296-a388-f5aeaa97722f)

First output is the serialized message and second is bs58 encoded message.

So I try to add the signatures too. Now I get a popup and can sign something. But the signature is just not the one I expect.

![Screenshot from 2023-11-08 08-19-25](https://github.com/phantom/docs/assets/14315050/50c247ca-b206-4b4d-b250-411e0a5113fe)

Again here the first output is the serialized transaction and the second is its base 58 encoded string. I get the signature with the following code:
`(await solana.request({
      method: 'signTransaction',
      params: {
        message: encoded_message,
      }
    })).signatures[0];`

It is not a signature of the transaction and I cannot verify it and I cannot send it.

### Phantom Version

23.15.0 chrome extension

### Is there an existing discussion for this?

- [X] I have searched the existing discussions

**Comment by adamdelphantom** [comment]
Hey @Virus-Axel thank you for writing in! And cool project!

Do you have some code available that I could take a look at?

**Reply by Virus-Axel** [reply]
Thank you for your answer, The code that is relevent from [https://github.com/Virus-Axel/godot-solana-sdk/blob/master/phantom/phantom.cpp](https://github.com/Virus-Axel/godot-solana-sdk/blob/master/phantom/phantom.cpp) is the methods sign_message and store_encoded_message I think. The javascript code is hardcoded at the top of the file. These will be executed in an eval(). The connect_phantom method works fine.

**Reply by Virus-Axel** [reply]
Also FYI the input transaction is a transfer transaction:

1, // number of signatures
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // empty signature

1, 0, 1, // message header (1 signer, of which 0 is readonly, 1 readonly non signer)
3, // total accounts.

20, 57, 225, 120, 103, 96, 103, 29, 157, 138, 191, 101, 253, 133, 18, 113, 31, 250, 46, 118, 39, 190, 250, 94, 32, 59, 119, 220, 114, 139, 120, 189, // My phantom payer key
222, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 222 // The receiver account
0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, // The program ID

212, 10, 172, 122, 91, 224, 199, 87, 21, 106, 247, 127, 77, 2, 6, 50, 210, 177, 167, 70, 101, 234, 170, 62, 6, 219, 228, 155, 44, 52, 48, 141, // Latest blockhash

1, // Number of instructions
2, // PID index
2, // Two other accounts
0, 1, // Account indicies
12, // data size
2, 0, 0, 0, 10, 0, 0, 0, 0, 0, 0, 0 // data

I am curious to know if this is what phantom expects (as bs58 encoded string)

**Comment by Virus-Axel** [comment, marked as answer]
I found the reason for this here: [https://solana.stackexchange.com/questions/6623/how-to-disable-priority-fees-in-phantom/6632#6632](https://solana.stackexchange.com/questions/6623/how-to-disable-priority-fees-in-phantom/6632#6632). The solution would be to add ComputeUnitBudget instruction and a ComputeUnitLimit to the transaction. That is what wallets sign anyways (some conditions to it). It was also possible to partially sign to get around this phenomenon. I tried partially signing but seems not possible when there is only one signer.

I am glad this is resolved, and thank you for your time :)

**Reply by adamdelphantom** [reply]
Glad to hear you were able to resolve! Let us know if there's anything else we can help with.
